lib/guardian/phoenix/socket.ex

```diff
@@ -0,0 +1,137 @@
+  end
+  ```
+
+  If you want more control over the authentication of the connection, then you shoule `import Guardian.Phoenix.Socket` and use the `sign_in` function to authenticate.
+
+  defmodule MyApp.UserSocket do
+    use Phoenix.Socket
+    import Guardian.Phoenix.Socket
+
+    def connect(%{"guardian_token" => jwt } = params, socket) do
+      case sign_in(socket, jwt) do
+        {:ok, authed_socket, guardian_params} ->
+          {:ok, authed_socket}
+        _ -> :error
+      end
+    end
+  end
+
+  If you want to authenticate on the join of a channel, you can import this module and use the sign_in function as normal.
+  """
+  defmacro __using__(opts) do
+    opts = Enum.into(opts, %{})
+    key = Map.get(opts, :key, :default)
+
+    quote do
+      import Guardian.Phoenix.Socket
+
+      def connect(%{ "guardian_token" => jwt } = params, socket) do
+        case sign_in(socket, jwt, params, key: unquote(key)) do
+          {:ok, authed_socket, _guardian_params} -> {:ok, authed_socket}
+          _ -> :error
+        end
+      end
+    end
+  end
+
+  def claims(socket, key \\ :default), do: current_claims(socket, key) # deprecated in 1.0
+
+  @doc """
+  Fetches the `claims` map that was encoded into the token.
+  """
+  def current_claims(socket, key \\ :default), do: socket.assigns[Guardian.Keys.claims_key(key)]
+
+  @doc """
+  Fetches the JWT that was provided for the initial authentication. This is provided as an encoded string.
+  """
+  def current_token(socket, key \\ :default), do: socket.assigns[Guardian.Keys.jwt_key(key)]
+
+  @doc """
+  Loads the resource from the serializer.
+  The resource is not cached onto the socket so using this function will load a fresh version of the resource each time it's called.
+  """
+  def current_resource(socket, key \\ :default) do
+    case current_claims(socket, key) do
+      nil -> nil
+      claims ->
+        case Guardian.serializer.from_token(claims["sub"]) do
+          {:ok, resource} -> resource
+          _ -> nil
+        end
+    end
+  end
+
+  @doc """
+  Boolean if the token is present or not to indicate an authenticated socket
+  """
+  def authenticated?(socket, key \\ :default) do
+    socket
+    |> current_token(key)
+    |> is_binary
+  end
+
+  def sign_in(_socket, nil), do: {:error, :no_token}
+  def sign_in(socket, jwt), do: sign_in(socket, jwt, %{})
+
+  @doc """
+  Sign into a socket. Takes a JWT and verifies it. If successful it caches the JWT and decoded claims onto the socket for future use.
+  """
+  def sign_in(socket, jwt, params, opts \\ []) do
+    key = Keyword.get(opts, :key, :default)
+
+    case Guardian.decode_and_verify(jwt, params) do
+      {:ok, claims} ->
+        case Guardian.serializer.from_token(Map.get(claims, "sub")) do
+          {:ok, resource} ->
+            authed_socket = socket
+            |> Phoenix.Socket.assign(Guardian.Keys.claims_key(key), claims)
+            |> Phoenix.Socket.assign(Guardian.Keys.jwt_key(key), jwt)
```

Comment:
What do you think about making `set_current_token` and `set_current_claims` helpers? These could be especially helpful from the tests, so you could do

```elixir
socket = Guardian.Phoenix.Socket.set_current_token(socket, "THE JWT")
```

instead of 
```elixir
key = Guardian.Keys.jwt_key
assigns = %{key => "THE JWT"}
socket = Map.put(socket, :assigns, assigns)
```

This would also more closely follow how it's currently being done in Guardian.Plug, and would allow for easier refactoring of where the value is being stored on the socket (like how we moved from using Plug.Conn.assign to Plug.Conn.put_private).

That said, I'm not sure if writing setters is "The Elixir Way", so I'm interested to hear your what you think.